### PR TITLE
fix: fix user data cache sync (SDKCF-3710)

### PR DESCRIPTION
**Pull request by imaginaris**
# Description
Anonymous user has now its own independent cache.
Last user cache is used only after app launch and before previously logged in user data is registered (and there was no logout  in previous app session)

## Links
SDKCF-3710

# Checklist
- [X] I have read the [contributing guidelines](../CONTRIBUTING.md).
- [X] I wrote/updated tests for new/changed code
- [X] I removed all sensitive data **before every commit**, including API endpoints and keys
- [X] I ran `fastlane ci` without errors

### CODE REVIEW

File: RInAppMessaging/Classes/CampaignDispatcher.swift [hunk 1]
```diff
@@ -44,9 +44,13 @@ internal class CampaignDispatcher: CampaignDispatcherType, TaskSchedulable {
     }
 
     func resetQueue() {
-        isDispatching = false
-        scheduledTask?.cancel()
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/CampaignDispatcher.swift:48]: Why were these moved into the async q?

File: RInAppMessaging/Classes/Protocols/Lockable.swift [hunk 2]
```diff
@@ -42,15 +45,15 @@ internal class LockableObject<T>: LockableResource {
         if shouldWait {
             dispatchGroup.wait()
         }
-        lockCount += 1
+        lockCount = lockCount + 1
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/Protocols/Lockable.swift:48]: How does this ensure thread safety?
Or is it because you've moved the decrement before the leave()?

File: RInAppMessaging/Classes/RInAppMessaging.swift [hunk 3]
```diff
@@ -1,3 +1,5 @@
+import Foundation
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/RInAppMessaging.swift:1]: Is that import required?

File: RInAppMessaging/Classes/Repositories/CampaignRepository.swift [hunk 4]
```diff
@@ -28,17 +28,21 @@ internal protocol CampaignRepositoryType: AnyObject, Lockable {
...
+    func loadCachedData(syncWithLastUserData: Bool)
+
+    /// Deletes cached data used to sync between users
+    func resetDataPersistence()
 }
 
 /// Repository to store campaigns retrieved from ping request.
 internal class CampaignRepository: CampaignRepositoryType {
 
+    static let lastUser = [UserIdentifier(type: .userId, identifier: "IAM.lastUser!@#")]
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/Repositories/CampaignRepository.swift:40]: Is the !@# just for uniqueness?

File: RInAppMessaging/Classes/Repositories/CampaignRepository.swift [hunk 5]
```diff
@@ -151,8 +159,6 @@ internal class CampaignRepository: CampaignRepositoryType {
     private func saveDataToCache(_ list: [Campaign]) {
         let user = preferenceRepository.getUserIdentifiers()
         userDataCache.cacheCampaignData(list, userIdentifiers: user)
-        if defaultCacheUser != user {
-            userDataCache.cacheCampaignData(list, userIdentifiers: defaultCacheUser)
-        }
+        userDataCache.cacheCampaignData(list, userIdentifiers: CampaignRepository.lastUser)
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/Repositories/CampaignRepository.swift:162]: Why is the same campaign list cached twice?

**Review by donnie-jp — APPROVED** [review]
LGTM - I left a few questions asking for clarifications but we can proceed to merge this so it can be QA'd.

File: RInAppMessaging/Classes/UserDataCache.swift [hunk 6]
```diff
@@ -2,6 +2,7 @@ internal protocol UserDataCacheable: AnyObject {
     func getUserData(identifiers: [UserIdentifier]) -> UserDataCacheContainer?
     func cacheCampaignData(_ data: [Campaign], userIdentifiers: [UserIdentifier])
     func cacheDisplayPermissionData(_ data: DisplayPermissionResponse, campaignID: String, userIdentifiers: [UserIdentifier])
+    func deleteUserData(identifiers: [UserIdentifier])
```
Comment (donnie-jp) [inline RInAppMessaging/Classes/UserDataCache.swift:5]: Where is the separate anonymous user cache that was mentioned in the PR description?

File: RInAppMessaging/Classes/CampaignDispatcher.swift (hunk 1 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/CampaignDispatcher.swift:48]: Because I think it wasn't really necessary to keep those outside, and now it's more thread safe.

File: RInAppMessaging/Classes/Protocols/Lockable.swift (hunk 2 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Protocols/Lockable.swift:48]: `@AtomicGetSet` doesn't work for incrementations. But actually this change doesn't help either. I'll have to fix that :)
Moving it before `leave()` might help though.

File: RInAppMessaging/Classes/RInAppMessaging.swift (hunk 3 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/RInAppMessaging.swift:1]: Sometimes when I clean the project I get a warning/error for a moment that I'm using `Foundation` features (`@objc`) without importing it. It works without that import in our app but it might not for others.

File: RInAppMessaging/Classes/Repositories/CampaignRepository.swift (hunk 4 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Repositories/CampaignRepository.swift:40]: Yes... 😬

File: RInAppMessaging/Classes/Repositories/CampaignRepository.swift (hunk 5 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/Repositories/CampaignRepository.swift:162]: It's basically the same approach as before but it doesn't store to anonymous user cache (to avoid userA -> logout -> userB data transfer)

File: RInAppMessaging/Classes/UserDataCache.swift (hunk 6 above)
Comment (imaginaris) [reply RInAppMessaging/Classes/UserDataCache.swift:5]: In CampaignRepository `lastUser`
